### PR TITLE
remove ignored avatars from PAL when they disconnect

### PR DESCRIPTION
**Pull request by davidkelly**
When you ignore someone, and they leave the domain, the entry for them remains in the PAL.  But, if they come back, there is a new entry for them.  The old entry is of no use, as unignoring is no longer possible.  So this PR will remove this entry when the avatar disconnects, since it is no longer of any use (unignore no longer will work once they leave the domain, even if they re-enter).

Test Plan:

Ignore an avatar.  Then, have that avatar go to another domain.  Refresh the PAL, and that entry should be gone.  Have the avatar come back.  Note that there is a new entry for them in the PAL, and they are not ignored.  

Before, you would have seen 2 entries for them, one the 'old' one (completely un-actionable since it refers to the old session), and the new one.  This simply removes the old one.

### CODE REVIEW

**Comment by hifi-gustavo** [comment]
Build console output for commit: 6c701bb0f0124002fb8e57c6a1cec8cfec1394ab available at: https://gist.github.com/7af759c690941b0c6f9eacb7140e885e<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR9404-6c701bb.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR9404-6c701bb.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 6c07a9aece6920794cc0d69a5f60d585911877ae available at: https://gist.github.com/7af759c690941b0c6f9eacb7140e885e<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR9404-6c07a9a.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR9404-6c07a9a.dmg

File: scripts/system/pal.js [hunk 1]
```diff
@@ -593,6 +596,8 @@ button.clicked.connect(onClicked);
 pal.visibleChanged.connect(onVisibleChanged);
 pal.closed.connect(off);
 Users.usernameFromIDReply.connect(usernameFromIDReply);
+Users.avatarDisconnected.connect(avatarDisconnected);
```
Comment (howard-stearns) [inline scripts/system/pal.js]: disconnect in scriptEnding callback?

File: libraries/networking/src/NodeList.cpp [hunk 2]
```diff
@@ -847,6 +847,17 @@ void NodeList::ignoreNodeBySessionID(const QUuid& nodeID, bool ignoreEnabled) {
     }
 }
 
+// removes this UUID from ignore and mute lists.  
+void NodeList::maintainIgnoreMuteSets(const QUuid& nodeID) {
```
Comment (howard-stearns) [inline libraries/networking/src/NodeList.cpp]: I feel weird using the word "maintain" to mean "stop maintaining" (i.e., to erase from the data).
Maybe removeFromIgnoreMuteSets()?

**Comment by howard-stearns** [comment]
**Additional test plan:**

See how fast you can go to another domain and come back. (Perhaps by going somewhere, then address-bar  back button to start the test. Then do address-bar forward-then-back very fast. (The address bar doesn't go away when you click forward and back, so you should be able to click pretty fast.)) See if the PAL removes the user correctly (and re-adds) in this case, too.

Try both ignored avatars going away (as above), and also muted avatars, and avatars that are neither ignored nor muted.

File: libraries/networking/src/NodeList.cpp (hunk 2 above)
Comment (davidkelly) [inline libraries/networking/src/NodeList.cpp]: yea -- I felt the same.  I'll use ^^

File: interface/src/avatar/AvatarManager.cpp [hunk 3]
```diff
@@ -261,6 +261,11 @@ void AvatarManager::handleRemovedAvatar(const AvatarSharedPointer& removedAvatar
     if (removalReason == KillAvatarReason::TheirAvatarEnteredYourBubble || removalReason == YourAvatarEnteredTheirBubble) {
         DependencyManager::get<NodeList>()->radiusIgnoreNodeBySessionID(avatar->getSessionUUID(), true);
     }
+    if (removalReason == KillAvatarReason::AvatarDisconnected) {
```
Comment (zfox23) [inline interface/src/avatar/AvatarManager.cpp]: Any reason this isn't an "else if()" ?

**Comment by hifi-gustavo** [comment]
Build console output for commit: 2460e89a574613ab615c05c0d2ba283a6d2a5d71 available at: https://gist.github.com/7af759c690941b0c6f9eacb7140e885e<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR9404-2460e89.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR9404-2460e89.dmg

File: interface/src/avatar/AvatarManager.cpp (hunk 3 above)
Comment (davidkelly) [inline interface/src/avatar/AvatarManager.cpp]: no reason not to - will change

File: libraries/script-engine/src/UsersScriptingInterface.h [hunk 4]
```diff
@@ -139,6 +139,12 @@ public slots:
     */
     void usernameFromIDReply(const QString& nodeID, const QString& username, const QString& machineFingerprint);
 
+    /**jsdoc
+     * Notifies scripts that a user has disconnected from the domain
+     * @function Users.avatar.avatarDisconnected
+     */
```
Comment (zfox23) [inline libraries/script-engine/src/UsersScriptingInterface.h]: For consistency, please add:
* `@param {nodeID} nodeID The node or session ID of the user who disconnected.`

**Comment by hifi-gustavo** [comment]
Build console output for commit: 9b1aaf3bfeb9a136a79396f75565eb02abd110a0 available at: https://gist.github.com/7af759c690941b0c6f9eacb7140e885e<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR9404-9b1aaf3.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR9404-9b1aaf3.dmg

**Comment by davidkelly** [comment]
please test this

**Comment by davidkelly** [comment]
test this please

**Comment by hifi-gustavo** [comment]
Build console output for commit: 9b1aaf3bfeb9a136a79396f75565eb02abd110a0 available at: https://gist.github.com/7af759c690941b0c6f9eacb7140e885e<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR9404-9b1aaf3.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR9404-9b1aaf3.dmg

**Pull request merged**